### PR TITLE
fix(calling): fix talking indicator

### LINKED ISSUE #766: bug(Voice): Improve voice/video indicator
**Issue:**
The voice and video indicator in calls don't always appear when they should and could use some improving. Sometimes when a user speaks it'll flash the indicator for a second and then disappear and then randomly flash again for a moment 

**Screenshot:**

https://github.com/user-attachments/assets/0d23e747-15fa-40b5-8247-91866e5566bd

### PR DESCRIPTION
**Pull request by Flemmli97**
### What this PR does 📖

- Fixes problems with talking indicator

### Which issue(s) this PR fixes 🔨

- Resolve #766

### CODE REVIEW

**Comment by phillsatellite** [comment]
Hey Kevin! @stavares843 and I noticed a small issue when we tested together

So Sara had a steady recording playing so talking indicator should remain on the entire time she is talking. 

On my side the indicator appeared fine but for Sara the indicator was not appearing. There was also a moment when we first joined the call when I was talking but talking indicator did not appear for me but that was only for a few minutes then started to appear 

This is my side and shows Sara's voice indicator working 
image


This is Sara's side and shows no indicator appearing 
image

**Comment by Flemmli97** [comment]
did you hear sara? and how loud was she?

**Comment by phillsatellite** [comment]
> did you hear sara? and how loud was she?

I could hear Sara and the volume was normal just like the other times

**Comment by Flemmli97** [comment]
> > did you hear sara? and how loud was she?
> 
> I could hear Sara and the volume was normal just like the other times

i added a log msg and also decreased the trigger threshold. can you check the volume in the log for incoming sound?
we could also just completely remove the treshold and just treat anything > 0 as talking

**Comment by phillsatellite** [comment]
Hey Kevin. Sara and I have retested and had some issues still 

I could see the indicator when Sara was talking (it would kinda flash a little. You can see in the video below) but whenever I would talk no indicator would appear around my own (Same with Sara she would never see her indicator) is that intended? 

Also when I would talk Sara also would not see the indicator for me 

https://github.com/user-attachments/assets/180052aa-4656-4809-9805-5a8c9a99f0b2

image

image

**Comment by Flemmli97** [comment]
> Hey Kevin. Sara and I have retested and had some issues still
> 
> I could see the indicator when Sara was talking (it would kinda flash a little. You can see in the video below) but whenever I would talk no indicator would appear around my own (Same with Sara she would never see her indicator) is that intended?
> 
> Also when I would talk Sara also would not see the indicator for me
> 
>  Screen_Recording_2024-10-29_at_11.36.22_AM.mov 
> image
> 
> image

ye. own talking indicator is unimplemented. also is the last screenshot from sara? when you were talking?

**Comment by phillsatellite** [comment]
> > Hey Kevin. Sara and I have retested and had some issues still
> > I could see the indicator when Sara was talking (it would kinda flash a little. You can see in the video below) but whenever I would talk no indicator would appear around my own (Same with Sara she would never see her indicator) is that intended?
> > Also when I would talk Sara also would not see the indicator for me
> > Screen_Recording_2024-10-29_at_11.36.22_AM.mov
> > image
> > image
> 
> ye. own talking indicator is unimplemented. also is the last screenshot from sara? when you were talking?

This is a screen recording from my side. I can get Sara to get a video of her side to show that there wasnt a indicator

**Comment by stavares843** [comment]
image


im talking and phil as well and i got no green indicator in either of the users

**Comment by Flemmli97** [comment]
> image
> 
> im talking and phil as well and i got no green indicator in either of the users

can you post the logs?

**Comment by Flemmli97** [comment]
also i added a talking indicator for oneself

**Comment by phillsatellite** [comment]
Tested again and it was technically working a little but you can see it just like flashes whenever Sara is talking 

https://github.com/user-attachments/assets/cc6a8fe9-0fea-4d38-ad18-6d1d79298784

**Comment by stavares843** [comment]
> also i added a talking indicator for oneself

phil and i were also not getting this

**Comment by Flemmli97** [comment]
can you try changing the `VOLUME_THRESHOLD` in voice.ts? i had it at 20. now its 0 so it will trigger at the slightest bit of sound.
also using chrome?

**Comment by stavares843** [comment]
yes we are using chrome

yes the VOLUME_THRESHOLD is already at 0, since its on the PR diff

**Comment by Flemmli97** [comment]
> yes we are using chrome
> 
> yes the VOLUME_THRESHOLD is already at 0, since its on the PR diff

can you post your logs when you hear phill talking?

**Comment by phillsatellite** [comment]
Hey Kevin. So Sara and I retested and its working better sorta but theres still an issue. 

So whenever we unmute the green indicator will appear even if there is zero sound and nothing being said the green talking indicator will still be on. This is the same for both Sara and I 

image

https://github.com/user-attachments/assets/229a82dd-acba-4db9-8b9e-fdc0e29ae6d3

**Comment by Flemmli97** [comment]
> Hey Kevin. So Sara and I retested and its working better sorta but theres still an issue.
> 
> So whenever we unmute the green indicator will appear even if there is zero sound and nothing being said the green talking indicator will still be on. This is the same for both Sara and I
> 
> image
> 
>  Screen_Recording_2024-11-01_at_1.16.40_PM.mov

ye i guess that why we should find a good sound threshold. judging by the logs sound is coming through. you just arent hearing anything cause its very quiet.
try setting `VOLUME_THRESHOLD` back to 20 (what i had before).
also try turning pc volume up to see if there is background noise etc.

**Comment by Flemmli97** [comment]
> > Hey Kevin. So Sara and I retested and its working better sorta but theres still an issue.
> > So whenever we unmute the green indicator will appear even if there is zero sound and nothing being said the green talking indicator will still be on. This is the same for both Sara and I
> > image
> > Screen_Recording_2024-11-01_at_1.16.40_PM.mov
> 
> ye i guess that why we should find a good sound threshold. judging by the logs sound is coming through. you just arent hearing anything cause its very quiet. try setting `VOLUME_THRESHOLD` back to 20 or maybe even higher (what i had before). also try turning pc volume up to see if there is background noise etc.

**Review by stavares843 — COMMENTED** [review]
phil and i retested again

browser - chrome

we updated to 20 as you mentioned

image


phil side

https://github.com/user-attachments/assets/b1664047-2445-444b-ba52-d981c430eb3c


my side

https://github.com/user-attachments/assets/7e25ec9c-0c23-41b5-ba4b-b84756539b70

so on both sides, phil saw my avatar always flashing and i saw his always flashing regardless if we were talking on that moment or not

**Review by InfamousVague — APPROVED** [review]

**Comment by Flemmli97** [comment]
> phil and i retested again
> 
> browser - chrome
> 
> we updated to 20 as you mentioned
> 
> image
> 
> phil side
> 
>  Screen_Recording_2024-11-04_at_2.58.02_PM.mov 
> my side
> 
>  Gravacao.do.ecra.2024-11-04.as.19.58.09.mov 
> so on both sides, phil saw my avatar always flashing and i saw his always flashing regardless if we were talking on that moment or not

ye i can hear audio incoming on phils side. so it is working as intended. what we could do to stop the flashing is: 
if user is "green" increase the delay for audio check so it leaves a bigger buffer when talking before turning to not talking. at the same time keeping a faster response when someone is not talking and starts talking

File: src/lib/media/Voice.ts [hunk 1]
```diff
@@ -80,6 +80,55 @@ export type CallUpdater = {
...
+        return dataArray.reduce((prev, value) => (prev && prev > value ? prev : value))
+    }
+
+    function updateMeta(did: string) {
+        let muted = stream.getAudioTracks().find(track => !track.enabled || track.readyState === "ended") !== undefined
+        let speaking = false
+        let user = Store.getUser(did)
+        let current = get(user)
+        let vol = volume()
+        console.log("incoming volume ", vol)
```
Comment (lgmarchi) [inline src/lib/media/Voice.ts:106]: Let's change it to our Logger.

File: src/lib/media/Voice.ts [hunk 2]
```diff
@@ -80,6 +80,55 @@ export type CallUpdater = {
...
+        if (!muted && vol > VOLUME_THRESHOLD) {
```
Comment (lgmarchi) [inline src/lib/media/Voice.ts:107]: Here I think we can do like this: 

```ts
speaking = !muted && vol > VOLUME_THRESHOLD

**Review by lgmarchi — APPROVED** [review]
Code looks good, I just left 2 suggestions.

File: src/lib/media/Voice.ts (hunk 1 above)
Comment (Flemmli97) [reply src/lib/media/Voice.ts:106]: see below

File: src/lib/media/Voice.ts (hunk 2 above)
Comment (Flemmli97) [reply src/lib/media/Voice.ts:107]: its atm done this way because for logging testing

Comment (lgmarchi) [reply src/lib/media/Voice.ts:107]: But what is difference for log or not on this conditional?

**Comment by Flemmli97** [comment]
closing in favor of #817

**Pull request closed without merging**